### ISSUE TITLE
support json output to allow for nicer interop with jest snapshot testing

### ISSUE THREAD

**Issue by conorhastings** [developit/preact-render-to-string#16]
Right now using preact-render-to-string directly you end up with double encoding in the snapshot output 
<img width="919" alt="screen shot 2016-12-04 at 8 02 07 pm" src="https://cloud.githubusercontent.com/assets/8263298/20870862/22e77eae-ba5d-11e6-9865-c9bd32e472e5.png">

see this twitter thread -- https://twitter.com/stillconor/status/805567906310451200 -- for more details.

cc @developit , I'd be interesting in taking this on if you could provide high level guidance on how this would would be worked on.

**Comment by developit** [comment]
@conorhastings That would be awesome - I've been pondering the best approach myself.  Right now the renderer doesn't use an intermediary, so it's straight to string concat.  I think the simplest approach would be to add a `{ json:true }` option added that builds up objects instead of strings.  It might end up being quite a few branches in the code though, not sure what to do about that.

**Comment by ruyadorno** [comment]
I have been doing just that in my current project using [jest-serializer-html-string](https://github.com/ruyadorno/jest-serializer-html-string) to serialize the html that gets rid of the double encoding

if that interests anyone I can put together a small boilerplate but should be fairly straightforward, just @developit [example](https://twitter.com/_developit/status/805569608782585856) + **jest** snapshotSerializer config

**Comment by developit** [comment]
that would be very useful @ruyadorno

**Comment by paranoidjk** [comment]
@ruyadorno  great job 👍 
our project(https://github.com/ant-design/ant-design-mobile) use jest snapshot test, if preact can works with jest snapshot test, then we will be more confident to use preact replace react

**Comment by ruyadorno** [comment]
I know you all have seen it in twitter but I'll leave the boilerplate link here for future reference:

https://github.com/ruyadorno/preact-jest-snapshot-test-boilerplate

**Comment by paranoidjk** [comment]
@ruyadorno @developit seems like the render result is consistent, but  html string is not serialized as the same way of enzyme-to-json
https://github.com/ant-design/ant-design-mobile/issues/1304

**Comment by paranoidjk** [comment]
@ruyadorno 
as https://github.com/ant-design/ant-design-mobile/pull/1306 ci result shownd,
The problem is [preact-render-to-string](https://github.com/developit/preact-render-to-string) not place html attribute at new line, so the react snapshot not be consistent with preact snapshot, I will create a PR to fix this.

**Comment by paranoidjk** [comment]
the difference is whether **put each prop on a separate line**.
from https://github.com/adriantoine/enzyme-to-json#user-content-serializer:

>This is inspired by jest-serializer-enzyme, I first added a note to jest-serializer-enzyme but I then realised that the output is different, so it is not retro compatible with enzyme-to-json because it's using Enzyme debug helper which doesn't put each prop on a separate line.

**Comment by developit** [comment]
@paranoidjk ah - you probably want to use the JSX string renderer:

```js
import preactRender from 'preact-render-to-string/jsx';

preactRender(<foo a="b" c="d">bar</foo>)
/*
<foo
  a="b"
  c="d"
>
  bar
</foo>
*/
```

**Comment by ruyadorno** [comment]
@paranoidjk yeah, even though it might be valid html, it's def not the standard way html is usually rendered, I'd rather stick with the default options from the [html beautifier](https://github.com/maxogden/commonjs-html-prettyprinter) within [jest-serializer-html-string](https://github.com/ruyadorno/jest-serializer-html-string) as that would make more sense for most people (including me).

**Comment by marvinhagemeister** [comment]
@ruyadorno [jest-serializer-html-string](https://github.com/ruyadorno/jest-serializer-html-string) doesn't seem to match jest's rendering behaviour. Jest is more like the jsx renderer, which has the advantage of clearer diffs on attribute changes. Because each attribute is a single line it's much easier to see which ones are removed and added when multiple attribute changes are made for a single component.

I love your package though, keep up the good work 👍

**Comment by developit** [comment]
Trying to think of the slimmest way to produce JSON output from the string renderer.  It's the same flow, just seems awkward to have to have that many branches.

**Comment by whogood** [comment]
@ruyadorno thanks for sharing your solution

**Comment by nathancahill** [comment]
Just came across this thread, my fork of [preact-render-to-string](https://github.com/developit/preact-render-to-string) renders to JSON: [preact-render-to-json](http://github.com/nathancahill/preact-render-to-json).

**Comment by ascorbic** [comment]
@nathancahill Posting this here, as you don't have issues enabled on your repo: preact-render-to-json doesn't work with hooks, because of developit/preact#1373
